webapp/tests/karma/ts/components/search-bar/search-bar.component.spec.ts

```diff
@@ -152,4 +200,171 @@ describe('Search Bar Component', () => {
+      expect(setSnackbarContentSpy.calledWith('some text')).to.be.true;
+      expect(telemetryService.record.calledWith('search_by_barcode:not_supported'));
+    });
+
+    it('should return false if user does not have permission', async () => {
+      sessionService.isAdmin.returns(false);
+      authService.has.resolves(false);
+      sinon.resetHistory();
+
+      await component.ngAfterViewInit();
+
+      expect(component.isBarcodeScannerAvailable).to.be.false;
+      expect(sessionService.isAdmin.calledOnce).to.be.true;
+      expect(authService.has.calledOnce).to.be.true;
+      expect(authService.has.args[0]).to.have.members([ CAN_USE_BARCODE_SCANNER ]);
+    });
+
+    it('should return false if user is admin', async () => {
+      sessionService.isAdmin.returns(true);
+      authService.has.resolves(true);
+      sinon.resetHistory();
+
+      await component.ngAfterViewInit();
+
+      expect(component.isBarcodeScannerAvailable).to.be.false;
+      expect(sessionService.isAdmin.calledOnce).to.be.true;
+    });
+
+    it('should return false if BarcodeDetector is not supported', async () => {
+      sessionService.isAdmin.returns(false);
+      authService.has.resolves(true);
+      translateService.instant.returns('some text');
+      const setSnackbarContentSpy = sinon.spy(GlobalActions.prototype, 'setSnackbarContent');
+      sinon.resetHistory();
+      component.windowRef = {};
+
+      await component.ngAfterViewInit();
+
+      expect(component.isBarcodeScannerAvailable).to.be.false;
+      expect(translateService.instant.calledWith('barcode_scanner.warning.not_supported')).to.be.true;
+      expect(setSnackbarContentSpy.calledWith('some text')).to.be.true;
+      expect(translateService.instant.calledWith('barcode_scanner.warning.not_supported')).to.be.true;
+      expect(feedbackService.submit.calledWith('some text')).to.be.true;
+      expect(setSnackbarContentSpy.calledWith('some text')).to.be.true;
+      expect(telemetryService.record.calledWith('search_by_barcode:not_supported'));
+    });
+  });
+
+  describe('Scan barcodes', () => {
+    it('should scan barcode and trigger search', fakeAsync(async () => {
+      sessionService.isAdmin.returns(false);
+      authService.has.resolves(true);
+      const imageHolder = { addEventListener: sinon.stub() };
+      const createElementStub = sinon.stub(documentRef.defaultView.document, 'createElement');
+      createElementStub.returns(imageHolder);
+      detectStub.resolves([{ rawValue: '1234' }]);
+      sinon.resetHistory();
+
+      await component.ngAfterViewInit();
+
+      expect(getSupportedFormatsStub.calledOnce).to.be.true;
+      expect(imageHolder.addEventListener.calledOnce).to.be.true;
+      expect(imageHolder.addEventListener.args[0][0]).to.equal('load');
+
+      const eventCallback = imageHolder.addEventListener.args[0][1];
+      eventCallback();
+      flush();
+
+      expect(telemetryService.record.calledWith('search_by_barcode:scan')).to.be.true;
```

Comment:
The only telemetry code we can't evaluate at this level of test is `search_by_barcode:open`. Am I right? We should add the evaluation in the e2e test.

Reply:
I found how to unit test it! Thanks